Added net7 support.

Haplois opened the pull request.

Haplois committed:
Added net7 support.

Haplois committed:
Merge branch 'main' into dev/haplois/net67

Haplois committed:
Warnings fixed. More to fix.

Haplois committed:
Merge branch 'main' into dev/haplois/net67

Haplois committed:
Warnings fixed.

nohwnd commented on .editorconfig:
```diff
@@ -312,8 +312,8 @@ csharp_prefer_simple_default_expression = true:suggestion
 csharp_style_prefer_local_over_anonymous_function = false:suggestion # not default, default is true, avoided because we use anonymous functions in multiple places and it does not make the code clearer
-csharp_style_prefer_index_operator = true:suggestion
-csharp_style_prefer_range_operator = true:suggestion
+csharp_style_prefer_index_operator = false:suggestion # not default, default is true, disabled because Index operator is not supported in all the TFMs we support.
```
Do you know which tfms don't support it? It would be worth mentioning it here so we don't have to investigate the motifvation for this comment when we move on to next version of netfx.

Haplois replied:
Added supported versions.

nohwnd commented on src/Microsoft.TestPlatform.CoreUtilities/PublicAPI/net7.0/PublicAPI.Shipped.txt:
```diff
@@ -0,0 +1,13 @@
+﻿#nullable enable
```
Is this file identical to the one for net6.0? Can we just rename the other file and include the same file for both tfms in the csproj? I don't think these files will grow apart in the future.

nohwnd replied:
https://github.com/microsoft/vstest/blob/main/src/Microsoft.TestPlatform.CoreUtilities/PublicAPI/net6.0/PublicAPI.Shipped.txt They seem identical on first look.

nohwnd commented on src/Microsoft.TestPlatform.CoreUtilities/ValidateArg.cs:
```diff
@@ -34,7 +34,7 @@ public static class ValidateArg
     /// </returns>
     [DebuggerStepThrough]
-    public static T NotNull<T>([ValidatedNotNull] T? arg, string parameterName)
+    public static T NotNull<T>([NotNull] T? arg, string parameterName)
```
What is the impact of this change?

Haplois replied:
No impact. That also works on net7.

nohwnd replied:
I was looking for motivation of this change. Why is this better than the attribute we had?

Evangelink replied:
It's shipped inside the framework starting with .net 5 or 6 while the other one has to be manually defined.

Haplois replied:
Attribute we had didn't suppress the warnings and we already had this so in the cases it suppressed it was just a duplicate.

nohwnd commented on src/Microsoft.TestPlatform.ObjectModel/PublicAPI/net7.0/PublicAPI.Shipped.txt:
```diff
@@ -0,0 +1,4 @@
+﻿#nullable enable
```
Same question as above, is this just a copy of net6.0 file?

Haplois replied:
no net6 here.

nohwnd commented on src/Microsoft.TestPlatform.PlatformAbstractions/PublicAPI/net7.0/PublicAPI.Shipped.txt:
```diff
@@ -0,0 +1,17 @@
+﻿#nullable enable
```
And here as well. Can this net6.0 file be reused here?

nohwnd commented on src/Microsoft.TestPlatform.PlatformAbstractions/netstandard/System/ProcessHelper.cs:
```diff
@@ -93,7 +93,7 @@ public void WaitForProcessExit(object? process)
     }
 
-    public IntPtr GetProcessHandle(int processId)
+    public nint GetProcessHandle(int processId)
```
did not even know that nint exists, cool.

Haplois committed:
PR fixes.

Evangelink commented on src/DataCollectors/Microsoft.TestPlatform.Extensions.EventLogCollector/EventLogDataCollector.cs:
```diff
@@ -566,16 +566,17 @@ private void ConfigureMaxEntries(CollectorNameValueConfigurationManager collecto
         }
 
-        if (!eventLogContainerFound)
+        if (eventLogContainerFound == false)
```
Previous code was ok to me. Usually when I see `== false`, I expect some nullable bool which isn't the case here.

Evangelink commented on src/DataCollectors/Microsoft.TestPlatform.Extensions.EventLogCollector/Microsoft.TestPlatform.Extensions.EventLogCollector.csproj:
```diff
@@ -6,18 +6,12 @@
-    <Reference Include="System.Xml" />
-  </ItemGroup>
-  <ItemGroup Condition="'$(TargetFramework)' == 'net6.0'">
+  <ItemGroup Condition="'$(TargetFramework)' == 'net7.0'">
```
Could we use `'$(DotNetBuildFromSource)' == 'true'` instead? This would avoid us to update the TFM every time.

Haplois replied:
No, not now. In a separate PR, I will remove most of the sourcebuild customizations and use the same targets. That change is a bit bigger, so duplicates for now.

Evangelink commented on src/Microsoft.TestPlatform.Common/DataCollection/DataCollectionAttachmentManager.cs:
```diff
@@ -290,7 +290,12 @@ private void AddNewFileTransfer(FileTransferInformation fileTransferInfo, AsyncC
                         uri,
                         friendlyName,
-                        Guid.Parse(testCaseId));
+#if NET7_0_OR_GREATER
+                        Guid.Parse(testCaseId, CultureInfo.InvariantCulture)
+#else
+                        Guid.Parse(testCaseId)
+#endif
```
Not sure if that's way better but an alternative is

```suggestion
                        Guid.Parse(testCaseId
#if NET7_0_OR_GREATER
                            , CultureInfo.InvariantCulture
#endif
                            )
```

Evangelink commented on src/Microsoft.TestPlatform.Extensions.BlameDataCollector/WindowsHangDumper.cs:
```diff
@@ -131,17 +131,17 @@ internal static void CollectDump(IProcessHelper processHelper, Process process,
                 PlatformArchitecture.ARM64 => "DumpMinitool.arm64.exe",
-                _ => null
+                _ => string.Empty
             };
 
-            if (dumpMinitoolName == null)
+            if (dumpMinitoolName == string.Empty)
```
Why this change from `null` to string empty?

Evangelink commented on src/Microsoft.TestPlatform.ObjectModel/TestCase.cs:
```diff
@@ -261,7 +262,13 @@ protected override void ProtectedSetPropertyValue(TestProperty property, object?
+                Id = value is Guid guid
+                    ? guid
+#if NET7_0_OR_GREATER
+                    : Guid.Parse((value as string)!, CultureInfo.InvariantCulture);
```
Ouch there is something fishy here. We should not be doing `as` cast to then bang, it's either nullable or we should use direct cast instead.

Evangelink commented on src/Microsoft.TestPlatform.PlatformAbstractions/PublicAPI/PublicAPI.Shipped.txt:
```diff
@@ -47,7 +47,7 @@ Microsoft.VisualStudio.TestPlatform.PlatformAbstractions.Interfaces.IProcessHelp
 Microsoft.VisualStudio.TestPlatform.PlatformAbstractions.Interfaces.IProcessHelper.GetCurrentProcessLocation() -> string!
 Microsoft.VisualStudio.TestPlatform.PlatformAbstractions.Interfaces.IProcessHelper.GetNativeDllDirectory() -> string!
-Microsoft.VisualStudio.TestPlatform.PlatformAbstractions.Interfaces.IProcessHelper.GetProcessHandle(int processId) -> System.IntPtr
+Microsoft.VisualStudio.TestPlatform.PlatformAbstractions.Interfaces.IProcessHelper.GetProcessHandle(int processId) -> nint
```
Isn't this going to break compat for old TFMs?

Haplois replied:
No, as long as we are using a recent compiler.

Evangelink commented on src/Microsoft.TestPlatform.Utilities/ClientUtilities.cs:
```diff
@@ -54,6 +54,8 @@ public static void FixRelativePathsInRunSettings(XmlDocument xmlDocument, string
 
     private static void AddRunSettingsDirectoryNode(XmlDocument doc, string path)
     {
+        ValidateArg.NotNull(doc.DocumentElement, nameof(doc.DocumentElement));
```
I'm not a big fan of this. This is going to throw `ArgumentNullException` while it's not the arg that is null but a field of it. IMO we should throw either `ArgumentException` or `InvalidOperationException` (I would go for the last one as I mentally associate it to "invalid state" but the first one is fine too).

Evangelink commented on src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs:
```diff
@@ -705,12 +705,8 @@ public static IEnumerable<string> FilterCompatibleSources(Architecture chosenPla
-
-        return targetPlatform == Architecture.X64 && !Is64BitOperatingSystem() || sourcePlatform != targetPlatform;
+        return sourcePlatform is not Architecture.Default and not Architecture.AnyCPU
+&& (targetPlatform == Architecture.X64 && !Is64BitOperatingSystem() || sourcePlatform != targetPlatform);
```
Nit: Indentation is broken here.

Evangelink commented on src/Microsoft.TestPlatform.Utilities/XmlUtilities.cs:
```diff
@@ -63,7 +63,7 @@ internal static bool IsValidNodeXmlValue(string xmlNodeValue, Func<string, bool>
 #else
         // fixing manually as we currently target to netcore 1.1 and we don't have default implementation for Escape functionality
-        var secureInnerXml = innerXml.IsNullOrEmpty() ? innerXml : innerXml.Replace("&", "&amp;").Replace("<", "&lt;").Replace(">", "&gt;").Replace("\"", "&quot;").Replace("'", "&apos;");
+        var secureInnerXml = (innerXml ?? string.Empty).Replace("&", "&amp;").Replace("<", "&lt;").Replace(">", "&gt;").Replace("\"", "&quot;").Replace("'", "&apos;");
```
You have changed the logic here. That's not a big deal but you are forcing unnecessary operations

Evangelink commented on src/SettingsMigrator/Migrator.cs:
```diff
@@ -153,65 +152,65 @@ private static void MigrateTestSettings(string oldTestSettingsPath, string newRu
 
         string? testTimeout = null;
-        if (testSettingsNodes.Timeout != null && testSettingsNodes.Timeout.Attributes[TestTimeoutAttributeName] != null)
+        if (testSettingsNodes?.Timeout?.Attributes?[TestTimeoutAttributeName] != null)
```
According to nullability above, `testSettingsNodes`  cannot be null

```suggestion
        if (testSettingsNodes.Timeout?.Attributes?[TestTimeoutAttributeName] != null)
```

Evangelink commented on src/SettingsMigrator/Migrator.cs:
```diff
@@ -153,65 +152,65 @@ private static void MigrateTestSettings(string oldTestSettingsPath, string newRu
 
         string? runTimeout = null;
-        if (testSettingsNodes.Timeout != null && testSettingsNodes.Timeout.Attributes[RunTimeoutAttributeName] != null)
+        if (testSettingsNodes?.Timeout?.Attributes?[RunTimeoutAttributeName] != null)
```
According to nullability above, `testSettingsNodes`  cannot be null

```suggestion
        if (testSettingsNodes.Timeout?.Attributes?[RunTimeoutAttributeName] != null)
```

Evangelink commented on src/SettingsMigrator/Migrator.cs:
```diff
@@ -153,65 +152,65 @@ private static void MigrateTestSettings(string oldTestSettingsPath, string newRu
 
         string? parallelTestCount = null;
-        if (testSettingsNodes.Execution != null && testSettingsNodes.Execution.Attributes[ParallelTestCountAttributeName] != null)
+        if (testSettingsNodes?.Execution?.Attributes?[ParallelTestCountAttributeName] != null)
```
According to nullability above, `testSettingsNodes`  cannot be null

```suggestion
        if (testSettingsNodes.Execution?.Attributes?[ParallelTestCountAttributeName] != null)
```

Evangelink commented on src/SettingsMigrator/Migrator.cs:
```diff
@@ -153,65 +152,65 @@ private static void MigrateTestSettings(string oldTestSettingsPath, string newRu
 
         string? hostProcessPlatform = null;
-        if (testSettingsNodes.Execution != null && testSettingsNodes.Execution.Attributes[HostProcessPlatformAttributeName] != null)
+        if (testSettingsNodes?.Execution?.Attributes?[HostProcessPlatformAttributeName] != null)
```
According to nullability above, `testSettingsNodes`  cannot be null

```suggestion
        if (testSettingsNodes.Execution?.Attributes?[HostProcessPlatformAttributeName] != null)
```

Evangelink commented on src/SettingsMigrator/Migrator.cs:
```diff
@@ -153,65 +152,65 @@ private static void MigrateTestSettings(string oldTestSettingsPath, string newRu
 
         // WebTestRunConfiguration node.
-        if (testSettingsNodes.WebSettings != null)
+        if (testSettingsNodes?.WebSettings != null)
```
According to nullability above, `testSettingsNodes`  cannot be null

```suggestion
        if (testSettingsNodes.WebSettings != null)
```

Evangelink commented on src/SettingsMigrator/Migrator.cs:
```diff
@@ -153,65 +152,65 @@ private static void MigrateTestSettings(string oldTestSettingsPath, string newRu
 
         // LegacySettings node.
-        AddLegacyNodes(testSettingsNodes, testTimeout, parallelTestCount, hostProcessPlatform, runSettingsXmlDoc);
+        AddLegacyNodes(testSettingsNodes!, testTimeout, parallelTestCount, hostProcessPlatform, runSettingsXmlDoc!);
```
According to nullability above, `testSettingsNodes`  cannot be null

```suggestion
        AddLegacyNodes(testSettingsNodes, testTimeout, parallelTestCount, hostProcessPlatform, runSettingsXmlDoc!);
```

Evangelink commented on src/SettingsMigrator/Migrator.cs:
```diff
@@ -153,65 +152,65 @@ private static void MigrateTestSettings(string oldTestSettingsPath, string newRu
 
         // DataCollectors node.
-        if (testSettingsNodes.Datacollectors != null && testSettingsNodes.Datacollectors.Count > 0)
+        if (testSettingsNodes?.Datacollectors != null && testSettingsNodes.Datacollectors.Count > 0)
```
According to nullability above, `testSettingsNodes`  cannot be null

```suggestion
        if (testSettingsNodes.Datacollectors != null && testSettingsNodes.Datacollectors.Count > 0)
```

Evangelink replied:
I believe you can also simplify further

```suggestion
        if (testSettingsNodes.Datacollectors?.Count > 0)
```

Evangelink commented on src/SettingsMigrator/Migrator.cs:
```diff
@@ -153,65 +152,65 @@ private static void MigrateTestSettings(string oldTestSettingsPath, string newRu
+        if (testSettingsNodes?.Timeout?.Attributes?[RunTimeoutAttributeName] != null)
         {
-            runTimeout = testSettingsNodes.Timeout.Attributes[RunTimeoutAttributeName].Value;
+            runTimeout = testSettingsNodes?.Timeout?.Attributes?[RunTimeoutAttributeName]?.Value;
```
You checked for null above so you don't need the various conditional access.

Haplois replied:
I do, it causes might be null warnings.

Evangelink commented on src/SettingsMigrator/Migrator.cs:
```diff
@@ -153,65 +152,65 @@ private static void MigrateTestSettings(string oldTestSettingsPath, string newRu
+        if (testSettingsNodes?.Execution?.Attributes?[ParallelTestCountAttributeName] != null)
         {
-            parallelTestCount = testSettingsNodes.Execution.Attributes[ParallelTestCountAttributeName].Value;
+            parallelTestCount = testSettingsNodes?.Execution?.Attributes?[ParallelTestCountAttributeName]?.Value;
```
You checked for null above so you don't need the various conditional access.

Haplois replied:
same, if I don't - causes might be null warnings.

Evangelink commented on src/SettingsMigrator/Migrator.cs:
```diff
@@ -153,65 +152,65 @@ private static void MigrateTestSettings(string oldTestSettingsPath, string newRu
+        if (testSettingsNodes?.Execution?.Attributes?[HostProcessPlatformAttributeName] != null)
         {
-            hostProcessPlatform = testSettingsNodes.Execution.Attributes[HostProcessPlatformAttributeName].Value;
+            hostProcessPlatform = testSettingsNodes?.Execution?.Attributes?[HostProcessPlatformAttributeName]?.Value;
```
You checked for null above so you don't need the various conditional access.

Evangelink commented on src/SettingsMigrator/Migrator.cs:
```diff
@@ -153,65 +152,65 @@ private static void MigrateTestSettings(string oldTestSettingsPath, string newRu
+        if (testSettingsNodes?.WebSettings != null)
         {
-            runSettingsXmlDoc.DocumentElement.AppendChild(runSettingsXmlDoc.ImportNode(testSettingsNodes.WebSettings, deep: true));
+            runSettingsXmlDoc?.DocumentElement?.AppendChild(runSettingsXmlDoc.ImportNode(testSettingsNodes.WebSettings, deep: true));
```
`runSettingsXmlDoc` cannot be null, please remove the conditional access.

Evangelink commented on src/SettingsMigrator/Migrator.cs:
```diff
@@ -153,65 +152,65 @@ private static void MigrateTestSettings(string oldTestSettingsPath, string newRu
         if (!runTimeout.IsNullOrEmpty())
         {
-            AddRunTimeoutNode(runTimeout, runSettingsXmlDoc);
+            AddRunTimeoutNode(runTimeout, runSettingsXmlDoc!);
```
Cannot be null

```suggestion
            AddRunTimeoutNode(runTimeout, runSettingsXmlDoc);
```

Evangelink commented on src/SettingsMigrator/Migrator.cs:
```diff
@@ -223,14 +222,14 @@ private static TestSettingsNodes ReadTestSettingsNodes(string testSettingsPath)
 
-            if (testSettingsNodes.Timeout != null && (testSettingsNodes.Timeout.Attributes[AgentNotRespondingTimeoutAttribute] != null ||
-                                                      testSettingsNodes.Timeout.Attributes[DeploymentTimeoutAttribute] != null || testSettingsNodes.Timeout.Attributes[ScriptTimeoutAttribute] != null))
+            if (testSettingsNodes.Timeout != null && (testSettingsNodes?.Timeout?.Attributes?[AgentNotRespondingTimeoutAttribute] != null ||
```
`testSettingsXmlDoc` cannot be null.

Evangelink commented on src/SettingsMigrator/Migrator.cs:
```diff
@@ -223,14 +222,14 @@ private static TestSettingsNodes ReadTestSettingsNodes(string testSettingsPath)
         }
 
-        return testSettingsNodes;
+        return testSettingsNodes!;
```
Cannot be null.

```suggestion
        return testSettingsNodes;
```

Evangelink commented on src/SettingsMigrator/Migrator.cs:
```diff
@@ -239,10 +238,10 @@ private static TestSettingsNodes ReadTestSettingsNodes(string testSettingsPath)
     private static void RemoveEmbeddedTestSettings(XmlDocument newXmlDoc)
     {
-        var testSettingsNode = newXmlDoc.DocumentElement.SelectSingleNode(@"/RunSettings/MSTest/SettingsFile");
+        var testSettingsNode = newXmlDoc?.DocumentElement?.SelectSingleNode(@"/RunSettings/MSTest/SettingsFile");
```
`newXmlDoc` cannot be null.

Evangelink replied:
Can you please review all nullability in this class.

Haplois replied:
I did. If I remove marks causes might be null warnings.

Evangelink commented on src/vstest.console/CommandLine/AssemblyMetadataProvider.cs:
```diff
@@ -58,7 +58,15 @@ public Architecture GetArchitecture(string assemblyPath)
             var assemblyName = AssemblyName.GetAssemblyName(assemblyPath);
-            archType = MapToArchitecture(assemblyName.ProcessorArchitecture, assemblyPath);
+
+            var processorArchitecture =
```
I believe this deserves a code explanation.

Evangelink commented on src/vstest.console/InProcessVsTestConsoleWrapper.cs:
```diff
@@ -81,7 +80,12 @@ public InProcessVsTestConsoleWrapper(ConsoleParameters consoleParameters)
         // Fill the parameters.
-        consoleParameters.ParentProcessId = Process.GetCurrentProcess().Id;
+        consoleParameters.ParentProcessId =
+#if NET7_0_OR_GREATER
```
I should have called this earlier but can you please add a helper for this so that the pragma is only in one place and all calling sites have a simpler code?

Haplois replied:
This change requires a lot of additional changes, (aka consolidating our fill-ins, making some internals visible etc. - we can do this, but in a different pr.)

nohwnd commented on shared/NullableAttributes.cs:
```diff
@@ -149,11 +146,3 @@ public MemberNotNullWhenAttribute(bool returnValue, params string[] members)
-/// Secret attribute that tells the CA1062 validate arguments rule that this method validates the argument is not null.
-/// </summary>
-[AttributeUsage(AttributeTargets.Parameter)]
-internal sealed class ValidatedNotNullAttribute : Attribute
```
Asked similar thing above. https://github.com/microsoft/vstest/pull/3944#discussion_r951173250

Evangelink replied:
Closing this one as I have replied on the other one. Looking at the usage gave me the answer.

Haplois commented on src/DataCollectors/Microsoft.TestPlatform.Extensions.EventLogCollector/EventLogDataCollector.cs:
```diff
@@ -566,16 +566,17 @@ private void ConfigureMaxEntries(CollectorNameValueConfigurationManager collecto
         }
     }
 
+    [return: NotNull]
```
It sees the return value as "might be null" without this.

Haplois committed:
Update src/Microsoft.TestPlatform.Common/DataCollection/DataCollectionAttachmentManager.cs

Co-authored-by: Amaury Levé <amauryleve@microsoft.com>

Haplois committed:
Update src/SettingsMigrator/Migrator.cs

Co-authored-by: Amaury Levé <amauryleve@microsoft.com>

Haplois committed:
Update src/SettingsMigrator/Migrator.cs

Co-authored-by: Amaury Levé <amauryleve@microsoft.com>

Haplois committed:
Update src/SettingsMigrator/Migrator.cs

Co-authored-by: Amaury Levé <amauryleve@microsoft.com>

Haplois committed:
Update src/SettingsMigrator/Migrator.cs

Co-authored-by: Amaury Levé <amauryleve@microsoft.com>

Haplois committed:
Update src/SettingsMigrator/Migrator.cs

Co-authored-by: Amaury Levé <amauryleve@microsoft.com>

Haplois committed:
Update src/SettingsMigrator/Migrator.cs

Co-authored-by: Amaury Levé <amauryleve@microsoft.com>

Haplois committed:
Update src/SettingsMigrator/Migrator.cs

Co-authored-by: Amaury Levé <amauryleve@microsoft.com>

Haplois committed:
Update src/SettingsMigrator/Migrator.cs

Co-authored-by: Amaury Levé <amauryleve@microsoft.com>

Haplois committed:
Update src/SettingsMigrator/Migrator.cs

Co-authored-by: Amaury Levé <amauryleve@microsoft.com>

Haplois commented on playground/MSTest1/MSTest1.csproj:
```diff
@@ -6,7 +6,7 @@
 
   <PropertyGroup>
-    <TargetFrameworks>$(TargetFrameworks);net472;net5.0</TargetFrameworks>
+    <TargetFrameworks>$(TargetFrameworks);net472;net5.0;net6.0;net7.0</TargetFrameworks>
```
Not really, we can remove it. I will.

Haplois commented on playground/MSTest1/MSTest1.csproj:
```diff
@@ -6,7 +6,7 @@
 
   <PropertyGroup>
-    <TargetFrameworks>$(TargetFrameworks);net472;net5.0</TargetFrameworks>
+    <TargetFrameworks>$(TargetFrameworks);net472;net5.0;net6.0;net7.0</TargetFrameworks>
```
No, not now. In a separate PR, I will remove most of the sourcebuild customizations and use the same targets. That change is a bit bigger, so duplicates for now.

Haplois committed:
PR fixes.

Evangelink committed:
Improve code

- Rework nullability
- Fix typos
- Remove unused directives
- Add pragma for directive used only under certain conditions

Evangelink committed:
Simplify nullability

Evangelink reviewed (approved):
Let's wait for tomorrow's discussion on netstandard before merging this one. There will be conflicts between the 2 PRs and netstandard is older so it would be better to fix conflicts here. If we decide to make more work on netstandard PR then we will merge this one first.

Evangelink commented:
@Haplois Ping me for a new review when conflicts are fixed.

Haplois committed:
Merge branch 'main' into dev/haplois/net67

Haplois committed:
Fixed build.

Haplois committed:
Update ConsoleLogger.cs

Haplois committed:
Update Microsoft.TestPlatform.CoreUtilities.csproj

Haplois committed:
Fixed build.
